### ISSUE TITLE
Float16 quantized DistilBERT model performs poorly

### ISSUE THREAD

**Issue by sayakpaul** [tensorflow/tensorflow#43719]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below): 2.3
- Python version:
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**

I used the DistilBERT model for a text classification task with the SST-2 dataset. Now, when I use the float16 quantization recipe to convert the trained model to TensorFlow Lite the performance drops significantly (90% validation accuracy to ~50%). Upon investigating more I found out the TensorFlow Lite float16 quantized model always outputs `nan`. I am not sure why this is the case. When I convert the original model using dynamic-range quantization the performance is retained greatly. 

**Describe the expected behavior**

The float16 quantized model should not return `nan`.

**Standalone code to reproduce the issue**
- [Model training and conversion notebook](https://github.com/sayakpaul/BERT-for-Mobile/blob/master/DistilBERT_SST-2_TPU.ipynb)
- [TensorFlow Lite model evaluationn notebook](https://colab.research.google.com/gist/sayakpaul/705d53e72f0f5722461f892150fde8b0/evaluation_sst-2_distilbert.ipynb)

This is a part of a project called [BERT for Mobile](https://github.com/sayakpaul/BERT-for-Mobile/) where I plan to compare two different BERT models designed specifically for mobile deployments - DistilBERT and MobileBERT. Currently, I have only taken text classification as the NLP task. I plan to also include question answering in the future. 

Cc: @MeghnaNatraj @khanhlvg

**Comment by sayakpaul** [comment]
Hi. Any updates on this?

**Comment by MeghnaNatraj** [comment]
@sayakpaul we're looking into this now, will provide an update as soon as we have one.

**Comment by sayakpaul** [comment]
Thanks, @MeghnaNatraj. If you'd need the `SavedModel` let me know.

**Comment by talumbau** [comment]
Hi,

I tried to reproduce this but was not successful. I executed the two models in a Python script like this:

```
 import numpy as np

  # Load the TFLite model and allocate tensors.
  fp32_model = "./distilbert_fp32.tflite"
  interpreterf32 = Interpreter(model_path=fp32_model)
  interpreterf32.allocate_tensors()
  fp16_model = "./distilbert_fp16.tflite"
  interpreterf16 = Interpreter(model_path=fp16_model)
  interpreterf16.allocate_tensors()

  # Get input and output tensors.
  input_details = interpreterf32.get_input_details()
  print(input_details)
  output_details = interpreterf32.get_output_details()

  # Test the model on random input data.
  input_shape0 = input_details[0]['shape']
  input_shape1 = input_details[1]['shape']
  input_data0 = np.array(np.random.random_sample(input_shape0), dtype=np.int32)
  input_data1 = np.array(np.random.random_sample(input_shape1), dtype=np.int32)

  interpreterf32.set_tensor(input_details[0]['index'], input_data0)
  interpreterf32.set_tensor(input_details[1]['index'], input_data1)
  interpreterf32.invoke()
  output_dataf32 = interpreterf32.get_tensor(output_details[0]['index'])
  print(output_dataf32)

  interpreterf16.set_tensor(input_details[0]['index'], input_data0)
  interpreterf16.set_tensor(input_details[1]['index'], input_data1)
  interpreterf16.invoke()
  output_dataf16 = interpreterf16.get_tensor(output_details[0]['index'])
  print(output_dataf16)

```

but when I executed, I got an expected output (a float value in each case, fp16 quantized model producing a value that is similar but not exactly the same as the fp32 model):

```
INFO: Created TensorFlow Lite delegate for select TF ops.
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

[{'name': 'serving_default_bert_input_ids:0', 'index': 0, 'shape': array([  1, 128], dtype=int32), 'shape_signature': array([ -1, 128], dtype=int32), 'dtype': <class 'numpy.int32'>, 'quantization': (0.0, 0), 'quantization_parameters': {'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}, 'sparsity_parameters': {}}, {'name': 'serving_default_bert_input_masks:0', 'index': 1, 'shape': array([  1, 128], dtype=int32), 'shape_signature': array([ -1, 128], dtype=int32), 'dtype': <class 'numpy.int32'>, 'quantization': (0.0, 0), 'quantization_parameters': {'scales': array([], dtype=float32), 'zero_points': array([], dtype=int32), 'quantized_dimension': 0}, 'sparsity_parameters': {}}]
[[0.4656006]]
[[0.46526226]]
```

I'm not sure how to proceed. Is it easy for you to attempt converting the model again to TFL using the latest code?

**Comment by talumbau** [comment]
Also, please confirm that if the problem is reproducing, I should see `Nan` as the output. thanks!

**Comment by sayakpaul** [comment]
Thanks for looking into this, @talumbau. Did you use the same notebooks (linked in the issue above) to generate the fp32 and fp16 models? Also, which TensorFlow version did you use?

**Comment by talumbau** [comment]
My colleague generated the graphs using your notebooks. I believe it was TF 2.3. Over to @khanhlvg to confirm.

**Comment by sayakpaul** [comment]
I see. Could it be because you used random data as the inputs?

I think it's probably best to use [this notebook](https://colab.research.google.com/gist/sayakpaul/705d53e72f0f5722461f892150fde8b0/evaluation_sst-2_distilbert.ipynb) if you want reproduce the results.

**Comment by talumbau** [comment]
Is it your hypothesis that an input to the fp16 model produces Nans or do you have an example of an input to the fp16 model that causes a Nan? If it is the latter, can you provide such an input?

**Comment by sayakpaul** [comment]
It's the latter. You can find the notebook from my previous comment to reproduce the same.

**Comment by talumbau** [comment]
I am able to reproduce and am attempting to debug now.

**Comment by talumbau** [comment]
It appears that very large values are stored in one of the float buffers of the float32 model. To be precise, Tensor at index 74 in the float32 DistilBERT model is filled with all values `1000000015047466219876688855040.0` (1.0000000150474662e+30). This exceeds the range of the float16 type. The corresponding buffer in the fp16 model (Tensor at index 749, which is the output Tensor of the Dequantize Op taking Tensor 74) has all values `inf` which appears to be the default way to resolve the cast of float32 to float16 when the value exceeds the max float16 value. The `inf` values are multiplied by `0`, with the result of `nan`, which then begins to propagate through the model. The damage is already done by the time we dequantize the fp16 nan to float32 nan, so one option might be to clamp to the max value of the float16 value when we are quantizing the weights tensors to float16. I will try converting the float32 Distilbert model with this option and re-run to see the impact on accuracy.

**Comment by sayakpaul** [comment]
Thank you for this insight. Do you think there might be similar reasons for [this issue](https://github.com/tensorflow/tensorflow/issues/44306)?

**Comment by talumbau** [comment]
For that issue, it doesn't look like float16 quantization is involved, so I suspect it's a different underlying problem.

**Comment by sayakpaul** [comment]
Yeah right. It involves dynamic-range. The broader problem is the same - terrible drop in the performance although I understand it may be a different conversion problem.

**Comment by khanhlvg** [comment]
@sayakpaul 
This commit should have fixed the fp16 issue and will be included in tf-nightly build tomorrow.
https://github.com/tensorflow/tensorflow/commit/612a5fb91ed6a6c229c5f4932307747699cabe90

**Comment by sayakpaul** [comment]
Thanks for letting me @khanhlvg. I think we found out an important numerical issue here and I am really glad the TFLite team took the time to investigate it. THANK yOU!

**Comment by talumbau** [comment]
Just a note to say that I converted the f32 DistilBERT model with https://github.com/tensorflow/tensorflow/commit/612a5fb91ed6a6c229c5f4932307747699cabe90 applied and re-ran the inputs. No `nan`s were in the output and the agreement was good, so I think it's appropriate to close. Thanks for finding and reporting!

**Comment by talumbau** [comment]
Yes correct. The output looked like this (ignore the debug printing):

```
index= 0
INFO: Created TensorFlow Lite delegate for select TF ops.
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.999979]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

Inference result with fp16 model: [0.9999791]
index= 1
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.04669181]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.04674396]
index= 2
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.99949145]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.99949086]
index= 3
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.99908924]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.9990898]
index= 4
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.00043762]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.00043774]
index= 5
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.99986213]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.999862]
index= 6
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.00101683]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.0010168]
index= 7
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here [657/1955] Inference result with fp32 model: [0.0096097]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.00960544]
index= 8
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.8588852]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.85885197]
index= 9
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.00107691]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.00107661]
index= 10
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.99995357]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.99995345]
index= 11
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.00048542]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.00048554]
index= 12
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.00036857]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.0003686]
index= 13
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.40758246]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.40804586]
index= 14
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.00037163]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.00037152]
index= 15
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.999975]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.999975]
index= 16
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.99706167]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.99706614]
index= 17
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.9996885]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.9996884]
index= 18
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.0014008]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.00140226]
index= 19
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 600 nodes with 6 partitions.

casting to float here
Inference result with fp32 model: [0.0035983]
INFO: TfLiteFlexDelegate delegate: 6 nodes delegated out of 712 nodes with 6 partitions.

casting to float here
Inference result with fp16 model: [0.00360444]
```
